clayer/pysertype.c

```diff
@@ -1063,76 +1071,124 @@ static inline uint32_t check_number_of_samples (long long n)
   return (uint32_t)n;
 }
 
-static bool readtake_pre (long long N, uint32_t *Nu32, dds_sample_info_t **info, ddspy_sample_container_t **container, ddspy_sample_container_t ***rcontainer)
+static PyObject *readtake_post (int32_t sts, collector_state_t *state)
 {
-  if (!(*Nu32 = check_number_of_samples (N)))
-    return false;
-
-  *info = dds_alloc (sizeof (**info) * *Nu32);
-  *container = dds_alloc (sizeof (**container) * *Nu32);
-  *rcontainer = dds_alloc (sizeof (**rcontainer) * *Nu32);
+  if (sts < 0)
+    return PyLong_FromLong ((long)sts);
 
-  for (uint32_t i = 0; i < *Nu32; ++i) {
-    (*rcontainer)[i] = &(*container)[i];
-    (*container)[i].usample = NULL;
+  PyObject *list = PyList_New((Py_ssize_t)state->count);
+  for (size_t i = 0; i < state->count; ++i)
+  {
+    PyObject *sampleinfo = get_sampleinfo_pyobject(&state->sample_infos[i]);
+    PyObject *item = Py_BuildValue("(y#O)", 
+                                   state->containers[i].usample,
+                                   (Py_ssize_t)state->containers[i].usample_size,
+                                   sampleinfo);
+    PyList_SetItem(list, (Py_ssize_t)i, item); // steals ref
+    Py_DECREF(sampleinfo);
+    dds_free(state->containers[i].usample);
   }
-  return true;
+
+  dds_free(state->containers);
+  dds_free(state->sample_infos);
+
+  return list;
 }
 
-static PyObject *readtake_post (int32_t sts, dds_sample_info_t *info, ddspy_sample_container_t *container, ddspy_sample_container_t **rcontainer)
+dds_return_t collector_callback_fn(
+  void *arg,
+  const dds_sample_info_t *info,
+  const struct ddsi_sertype *sertype,
+  struct ddsi_serdata *serdata)
 {
-  if (sts < 0)
-    return PyLong_FromLong ((long)sts);
+  collector_state_t *state = (collector_state_t *)arg;
 
-  PyObject *list = PyList_New (sts);
-  for (int32_t i = 0; i < sts; ++i)
+  if (state->count >= state->max_samples)
+    return DDS_RETCODE_OK;
+
+  if (state->count >= state->capacity)
   {
-    PyObject *sampleinfo = get_sampleinfo_pyobject (&info[i]);
-    PyObject *item = Py_BuildValue ("(y#O)", container[i].usample, container[i].usample_size, sampleinfo);
-    PyList_SetItem (list, i, item); // steals ref
-    Py_DECREF (sampleinfo);
-    dds_free (container[i].usample);
+    // Grow allocation, this ensures amortized linear growth while keeping allocation calls minimal.
+    // Doubling gives exponential growth this makes adding N items only require log2(N) reallocations - efficient!
+    size_t new_capacity = state->capacity ? state->capacity * 2 : 8;
+    
+    void *new_containers = dds_realloc(state->containers, new_capacity * sizeof(ddspy_sample_container_t));
+    void *new_infos = dds_realloc(state->sample_infos, new_capacity * sizeof(dds_sample_info_t));
+
+    if (!new_containers || !new_infos)
+      return DDS_RETCODE_OUT_OF_RESOURCES;
```

Comment:
This I think is fine: it simply means the take/read operation stops early, and then with the modified condition in `readtake_post` it will return `DDS_RETCODE_OUT_OF_RESOURCES` if this is on the initial sample, or the data it did collect if it is on a later one